### PR TITLE
Correct handling of select-project dropdown based on number of projects and can create privileges

### LINKED ISSUE #477: Hide the add to project dropdown if no project
I stumbled on this for a second, probably indicating that it could be clearer. Usually if there isn't a choice we make a dropdown into a heading.

I have no projects, should I add to project or create a new? Both? Having no projects (but not realizing it atm since I create & delete them all the time) I found this odd:
screen shot 2017-10-09 at 2 55 32 pm

Hmm, no options, so why is it a drop down?
screen shot 2017-10-09 at 2 59 37 pm

Kind of clear, but is still a bit abnormal.
screen shot 2017-10-09 at 2 59 46 pm

I'm open to disagreement here, just pointing it out as we usually don't show a drop-down menu if there isn't actually an item to select.

@spadgett @jeff-phillips-18 @rhamilto

### LINKED ISSUE #494: Add custom msg for when there are no projects and user cannot create projects
This is for the `<select-projects>` component.

**No existing projects and cannot create projects**
image

Seems like a situation where it would be good to let the admin put a url or email address or something in the message.  We support custom messages and add it to the projects sidebar. We could show it here as well.

### PR DESCRIPTION
**Pull request by dtaylor113**
**Todo:**
- [x] Add all usecases to `<select-project>` component
- [x] Move `<select-project>` to top of form and add no-projects-cant-create empty state to *create-from-builder*
- [x] Add no-projects-cant-create empty state to *order-service*

Fixes #477
Fixes #494

**Latest Screenshots**
*Captions below screenshots*

image
---
image
---
image
---
image
---
image
---
image

### CODE REVIEW

**Comment by beanh66** [comment]
@dtaylor113 these look awesome, thanks for sharing all the screenshots. My only comment would be on the final use case (No projects and can't create a project, default message). I think the code snippet should be styled differently if possible. On the Homepage it is styled that way because of the dark background of the project panel, but on a white background I think we would show code with the light grey background and normal font color. Something like this:
 `oc adm new-project <projectname> --admin=<yourUsername>`

@jennyhaines can you confirm?

**Comment by jennyhaines** [comment]
@dtaylor113 @beanh66 - Yes, I agree with Colleen about styling this code snippet differently. Right now it's too stark/contrasty. I would actually use the same style that PatternFly uses for [inline code](http://www.patternfly.org/pattern-library/widgets/#cnt-remaining-chars). (Blue on light blue.) 

Here's an example of that style implemented in OpenShift Container Registry (circa June 2017) 
oscontainerregistry

**Comment by spadgett** [comment]
If you put it in a `<code>` tag, it should use this style. It's what we use elsewhere.

https://github.com/openshift/origin-web-console/blob/master/app/views/_cannot-create-project.html#L2

**Comment by beanh66** [comment]
Perfect, thanks for the update @jennyhaines!

**Comment by jennyhaines** [comment]
@dtaylor113 In that case, I would default to using the <code> tag that @spadgett mentioned, since you're using that elsewhere (rather than using the blue style in PF).

**Comment by dtaylor113** [comment]
@beanh66, ok changed to:

image

**Comment by dtaylor113** [comment]
@beanh66, changed no projects/empty state msg to "Services cannot be provisioned without a project."

**Comment by beanh66** [comment]
@dtaylor113 LGTM!

**Comment by spadgett** [comment]
@dtaylor113 you can test large project lists using this option with mock data:

https://github.com/openshift/origin-web-catalog/blob/master/app/config.js#L46-L47

File: src/components/select-project/selectProject.html [hunk 1]
```diff
@@ -1,9 +1,23 @@
-<ng-form name="$ctrl.forms.selectProjectForm">
+<span ng-if="$ctrl.noProjectsCantCreate" class="no-projects-cant-create">
+  <pf-empty-state config="$ctrl.noProjectsConfig"></pf-empty-state>
+  <div ng-if="!$ctrl.noProjectsCantCreateMessage">
+    A cluster admin can create a project for you by running the command:
+    <div class="code-block">
+      <code>oc adm <span class="command-arg">new-project</span> &lt;projectname&gt; <span class="command-arg">--admin={{$ctrl.user.metadata.name || '&lt;YourUsername&gt;'}}</span></code>
```
Comment (spadgett) [inline src/components/select-project/selectProject.html]: Is `$ctrl.user.metadata.name` ever set?

**Comment by dtaylor113** [comment]
> @dtaylor113 you can test large project lists using this option with mock data:

Hi @spadgett, I did have to make a one line change to get the large project list/filter to work.   I noticed that if you have a large project list and you can create projects, you do not get a change to create a project -the 'Create Project' option is not added to the dropdown.  So you can only select a project from the large list, you cannot create one.  Is this intentional?

**Comment by spadgett** [comment]
It's intentional for now. The problem is that if we added "Create Project" to the dropdown, it's not obvious you can search for other projects when you have large project lists.

File: src/components/select-project/selectProject.html (hunk 1 above)
Comment (dtaylor113) [reply src/components/select-project/selectProject.html]: Nice catch!  Thanks, fixed:
image

**Comment by dtaylor113** [comment]
> @dtaylor113 you can test large project lists using this option with mock data:

Fixed, tested

File: src/components/select-project/select-project.controller.ts [hunk 2]
```diff
@@ -95,6 +120,10 @@ export class SelectProjectController implements angular.IController {
     return this.projects && this.ctrl.selectedProject && !_.has(this.ctrl.selectedProject, 'metadata.uid');
   }
 
+  public canOnlyCreateProject(): boolean {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply src/components/select-project/select-project.controller.ts]: I call `canOnlyCreateProject()` three times from the HTML, I could replace each call with `(this.ctrl.numProjectChoices === 1 && this.ctrl.canCreate)`, or make canOnlyCreateProject a variable.  Though a helper method was warranted :-)

File: src/components/select-project/select-project.controller.ts [hunk 3]
```diff
@@ -214,12 +247,18 @@ export class SelectProjectController implements angular.IController {
...
       if (_.size(this.projects) === 1) {
         this.ctrl.selectedProject = createProject;
         this.onSelectProjectChange();
       }
+    } else if (_.size(this.projects) === 0) {
+      this.ctrl.noProjectsCantCreate = true;
+      this.$scope.$emit('no-projects-cannot-create');
     }
+
+    this.ctrl.numProjectChoices = _.size(this.projects);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply src/components/select-project/select-project.controller.ts]: Fixed, tested

**Comment by dtaylor113** [comment]
> the line is visible even when there are no other configuration parameters.
Will there ever be a usecase where `<select-project>` is the only config options?  To fix this I'd have to move the line out of the `<select-project>` component and move it to the 5 wizards (create-from-builder, deploy-image, etc..), and only show it if the wizard has other config params/settings.   -which I can do if we feel it is necessary.

**Comment by spadgett** [comment]
> Will there ever be a usecase where <select-project> is the only config options?

Yes. Not all services take parameters.

Maybe just add an optional flag to select-project saying there are no other fields?

**Comment by dtaylor113** [comment]
> the line is visible even when there are no other configuration parameters.

Fixed:
image

File: src/styles/order-service.less [hunk 4]
```diff
@@ -116,13 +116,25 @@
...
+    margin-left: -10px;
+  }
   h3 {
     line-height: 1.4;
     margin-top: 0;
     + .alert {
       margin-top: 20px;
     }
   }
+  .select-project-divider {
```
Comment (rhamilto) [inline src/styles/order-service.less]: Nit: alpha

File: src/styles/order-service.less [hunk 5]
```diff
@@ -116,13 +116,25 @@
...
+    border-bottom: 1px solid @color-pf-black-400;
+
```
Comment (rhamilto) [inline src/styles/order-service.less]: Nit:  unnecessary line

File: src/styles/order-service.less [hunk 6]
```diff
@@ -116,13 +116,25 @@
...
+    margin-top: 22px;
```
Comment (rhamilto) [inline src/styles/order-service.less]: Nit: alpha

File: src/components/select-project/selectProject.html [hunk 7]
```diff
@@ -1,9 +1,23 @@
-<ng-form name="$ctrl.forms.selectProjectForm">
+<span ng-if="$ctrl.noProjectsCantCreate" class="no-projects-cant-create">
+  <pf-empty-state config="$ctrl.noProjectsConfig"></pf-empty-state>
+  <div ng-if="!$ctrl.noProjectsCantCreateMessage">
```
Comment (rhamilto) [inline src/components/select-project/selectProject.html]: A &lt;p&gt; is a better choice here as it adds semantic meaning.

File: src/components/select-project/selectProject.html [hunk 8]
```diff
@@ -1,9 +1,23 @@
-<ng-form name="$ctrl.forms.selectProjectForm">
+<span ng-if="$ctrl.noProjectsCantCreate" class="no-projects-cant-create">
```
Comment (rhamilto) [inline src/components/select-project/selectProject.html]: &lt;div&gt; is probably a better choice here since this is a decent-sized block of markup.

**Comment by dtaylor113** [comment]
@rhamilto, thanks for review, I believe I have addressed all your issues.

File: src/styles/order-service.less [hunk 9]
```diff
@@ -116,13 +116,19 @@
...
+    margin-left: -10px;
+  }
   h3 {
     line-height: 1.4;
     margin-top: 0;
     + .alert {
       margin-top: 20px;
     }
   }
+  .no-projects-cant-create p, .no-projects-cant-create div  {
```
Comment (rhamilto) [inline src/styles/order-service.less]: Nit:  would be great to nest here:

```
.no-projects-cant-create {
  div,
  p {

```

**Comment by rhamilto** [comment]
The use of pf-empty-state here and https://github.com/openshift/origin-web-catalog/pull/501 gives me pause as this introduces a new empty state style to the console.  The console had an empty state pattern before pf-empty-state existed and the existing pattern is being used in all contexts that need it.  Adding pf-empty-state makes the aforementioned instances inconsistent with the rest of the console.  I understand the desire to use the official PatternFly components, but I think it's actually a negative in this case because it's introducing inconsistency in the UI.

**Comment by serenamarie125** [comment]
@rhamilto there are already inconsistencies.  See the associated design story identifying this issue https://patternfly.atlassian.net/browse/OSUX-155

**Comment by spadgett** [comment]
> there are already inconsistencies. See the associated design story identifying this issue https://patternfly.atlassian.net/browse/OSUX-155

2 of the 4 screenshots in that issue are from new designs in 3.7 that didn't account for what we had done elsewhere in the console. And I'd argue the other two from before 3.7 are more similar than different. This is mostly a problem we've introduced in 3.7. Adding yet another empty state style seems like a step backwards.

I'm not opposed to changing the style, but I really value consistency. I'd rather change them all to be consistent when we are able to.

**Comment by serenamarie125** [comment]
@spadgett I'm not arguing that we shouldn't make them consistent, just pointing out that the consistency already exists, and we are taking a look at it

**Comment by dtaylor113** [comment]
Hi @spadgett, which empty state would you like me to use in this case?  Could you please send a link to an example in the code base.  -thanks

**Comment by dtaylor113** [comment]
Hi @serenamarie125, @spadgett,  any chance of letting this PR merge and addressing all the empty state versions in a follow up PR when we address https://patternfly.atlassian.net/browse/OSUX-155?

**Comment by dtaylor113** [comment]
**Removed the icon from the empty state:**

image

File: src/components/order-service/order-service.controller.ts [hunk 10]
```diff
@@ -194,7 +200,7 @@ export class OrderServiceController implements angular.IController {
     this.updateBindability();
 
     this.validityWatcher = this.$scope.$watch("$ctrl.forms.orderConfigureForm.$valid", (isValid: any, lastValue: any) => {
-      this.configStep.valid = isValid;
+      this.configStep.valid = this.ctrl.noProjectsCantCreate === true ? false : isValid;
```
Comment (spadgett) [inline src/components/order-service/order-service.controller.ts]:
```js
this.configStep.valid = isValid && !this.ctrl.noProjectsCantCreate;
```

File: src/components/select-project/select-project.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply src/components/select-project/select-project.controller.ts]: I see now. I think this is fine.

**Comment by spadgett** [comment]
@dtaylor113 LGTM except for my one nit. @rhamilto had one or two comments pending as well

File: src/components/order-service/order-service.controller.ts (hunk 10 above)
Comment (dtaylor113) [reply src/components/order-service/order-service.controller.ts]: Fixed

File: src/styles/order-service.less (hunk 9 above)
Comment (dtaylor113) [reply src/styles/order-service.less]: Fixed

**Pull request merged**